### PR TITLE
Force with credentials

### PR DESCRIPTION
**Pull request by metromoxie**
@devd can you look at this one? In short, I think I understand the credential issue a lot better now, so hopefully this fixes the actual issue without introducing any others. I've also been more explicit in how it integrates with XHR.

### CODE REVIEW

**Comment by metromoxie** [comment]
Also, @annevk, I hope this clarifies this particular option's relation with the XHR processing model a bit (and also clarifies that it shouldn't interfere with the regular fetch() algorithm).

**Comment by annevk** [comment]
Regular resources do send cookies though.

**Comment by metromoxie** [comment]
What I meant by "excludes regular resource requests and fetch" is that the flag does nothing to regular resource requests and fetch. Indeed, the reason we don't need to do anything special for regular requests is that we are not modifying RFC 6266, so cookies are still sent on regular resource requests essentially based on the *physical origin* and not the suborigin (hence why we need the document to be cookie-averse). So regular resource requests have cookies attached across the physical origin, and there's nothing special with suborigins essentially.

**Comment by annevk** [comment]
So only a dying API is made special? Seems rather broken for a new feature.

**Comment by annevk** [comment]
But also, you state "and the later is unneeded because credentials are not sent by default, even to same-origin" which is false.

**Comment by metromoxie** [comment]
Ah, I see, so if this makes sense for withCredentials=true, then it ought to make sense for mode="same-origin". That makes sense to me. My aim is to make the change in behavior for requests and credentials act similarly across requests, and I was thinking XHR is a special case.

So would a solution here be to modify this CL so that it updates Fetch instead so that credentials are added if mode is same-origin and the request and url are same physical origin and the unsafe-with-credentials-flag is set?

**Comment by metromoxie** [comment]
(I suppose it would really be not changing the mode from "same-origin" and then letting the connection go w/out CORS if it's same physical origin, but I can work out those spec details if the approach makes sense.)

**Comment by annevk** [comment]
withCredentials=false is credentials mode "same-origin", true is "include". CORS uses "same-origin" by default, legacy stuff (`` without crossorigin) "include". But I'm not really sure what you're trying to do here.

**Comment by metromoxie** [comment]
The goal per @arturjanc's comments in #33 (and in particular I think [Artur's use case comment](https://github.com/w3c/webappsec-suborigins/issues/33#issuecomment-216776899) is very helpful) is to allow legacy clients to move to suborigins with primarily server-side changes.

In particular, @arturjanc points out that if a client uses XHR for same-origin requests, these would suddenly start failing because `withCredentials` hadn't been set, and since the credentials mode is then "same-origin", the request doesn't send the (now cross-origin) credentials. This would require a modification to the client, which in certain situations would be prohibitively expensive. @arturjanc's proposed solution is to force `withCredentials` to true if a special suborigin flag is set and the physical origins match.

My initial thought was that this is only necessary for XHR since the default for legacy resources is "include" anyway (so no client change is needed for credentials to be included), and for fetch() the default is no credentials for same origin or cross origin.

However, as you pointed out, if a user sets "same-origin" for fetch, then you have the same issue, where it would unexpectedly fail and require client changes as well. Would it make sense then to instead of special casing XHR make the change so that indicating a credentials mode of "same-origin" with the `'unsafe-with-credentials`' suborigin flag set the credentials flag to true if it's the same physical origin?

**Comment by annevk** [comment]
I see. The concern also applies to `` and everything with such an attribute. So yes, you want to patch Fetch's "same-origin" credentials mode. It should also be named after that rather than XHR. Just `unsafe-credentials` or longer `unsafe-same-origin-credentials` perhaps.

Is suborigins getting to the point where review would be helpful?

**Comment by metromoxie** [comment]
Great, this will make it feel less gross to me! I'll work on defining this patch to Fetch either Sunday or Monday.

No, I don't think reviewing it quite yet is a good idea. @devd and I are hammering out stuff still. Perhaps in a week or two, thanks!

**Comment by metromoxie** [comment]
@annevk and @devd, this is a first whack at Fetch patching (instead of XHR).

**Comment by metromoxie** [comment]
Forgot the most important part; actually checking for the flag! Will add shortly.

**Comment by metromoxie** [comment]
Added. You can also see the latest at https://metromoxie.github.io/webappsec-suborigins/.

File: index.bs [hunk 1]
```diff
@@ -782,28 +801,38 @@ spec:dom-ls; type:interface; text:Document
...
+     context will include <a>credentials</a>.
+
+     Specifically, update the step 2 of <a>HTTP-network-or-cache fetch</a> in
+     the Fetch spec [[!FETCH]] to read:
+
+     2. Let credentials flag be set if one of
+        * `request`'s credentials mode is "include"
+        * `request`'s credentials mode is "same-origin" and request's response
+          tainting is "basic"
+        * `request`'s credentials mode is "same-origin" and `request`'s
```
Comment (devd) [inline index.bs]: not sure I understand: why would the credentials mode be same-origin if making an XHR from a suborigin to same physical origin?

File: index.bs [hunk 2]
```diff
@@ -782,28 +801,38 @@ spec:dom-ls; type:interface; text:Document
...
+          <a>environment settings object</a> has the `suborigin unsafe
+          credentials` flag set and the request's <a>current url</a> is
+          <a>same-physical-origin</a> with <a>request origin</a>.
+
+     Note: This has several important, practical effects. If the <a>credentials
```
Comment (devd) [inline index.bs]: I wonder if worth clarifying that "and creds will be set even for requests to another sub-origin in the same physical origin"

File: index.bs [hunk 3]
```diff
@@ -609,20 +618,30 @@ spec:dom-ls; type:interface; text:Document
 
   </ol>
 
+  ## Comparing Physical Origins ## {#comparing-physical-origins}
+
+  Two suborigins are "the same physical origin" if, and only if, the
+  physical origin portion of the suborgiin are idential. In particular:
```
Comment (devd) [inline index.bs]: why was this whole section changed in this diff?

File: index.bs (hunk 2 above)
Comment (metromoxie) [inline index.bs]: Added

**Comment by metromoxie** [comment]
@annevk, I'm going to commit this as I believe this generally matches the conclusion from the earlier discussion, but I'd love your perspective on the update when you get a chance, and we can correct and polish it down the road.

**Pull request merged**